### PR TITLE
Use GetTransaction for unit-testing transaction based context

### LINKED ISSUE #49: Expose external transaction support for datastore methods
### Is your feature request related to a problem? Please describe.

To support transaction boundaries single transaction need to be created and should be pass to all the CRUD method invoke as a part this transaction.

Current implementation creates new transaction forever method invoke. 

Transaction boundaries can include multiple table and row updates, delete or new insert. 

### Describe the solution you'd like

Transaction can be part of tenant context and if tenant context has transaction then datastore should use it and avoid creating new transaction.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by krishnamiriyala**
Fixes #49

### CODE REVIEW

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 1]
```diff
@@ -41,22 +41,27 @@ func rollbackTx(t *testing.T, tx *gorm.DB) {
...
+func testTxCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, user1, user2 *AppUser) {
 	t.Helper()
 	assert := assert.New(t)
 	var err error
 
 	txFetcher := authorizer.SimpleTransactionFetcher{}
 
-	// Querying of previously inserted records should succeed
+	t.Log("Querying created records in single transaction")
+	tx, err := ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:52]: Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 2]
```diff
@@ -65,58 +70,56 @@ func testTxCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, myCok
 	user2.Email = "jahangir111@mail.com"
 	user2.EmailConfirmed = !user2.EmailConfirmed
 	user2.NumFollowers--
-
-	// TODO: Find, Update and Delete are still not supported due to appending where clauses,
-	// after each call to Find/Update/Delete methods
-
-	tx, err := ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
+	tx, err = ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:73]: Same

Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 3]
```diff
@@ -65,58 +70,56 @@ func testTxCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, myCok
...
+		err = ds.Find(txCtx, queryResult)
 		assert.NoError(err)
 		assert.Equal(record, queryResult)
 	}
+	assert.NoError(tx.Commit().Error)
 
+	t.Log("Verifying upsert and find in single transaction")
+	user1.NumFollowers++
+	user2.NumFollowers--
 	tx, err = ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:92]: Same

Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 4]
```diff
@@ -65,58 +70,56 @@ func testTxCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, myCok
...
-		err = ds.Find(ctx, queryResult)
+		err = ds.Find(txCtx, queryResult)
 		assert.NoError(err)
 		assert.Equal(record, queryResult)
 	}
+	assert.NoError(tx.Commit().Error)
 
-	// Deletion of existing records should not fail, and the records should no longer be found in the DB
+	t.Log("Verifying deletion and find in single transaction")
+	tx, err = ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:109]: Same

Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 5]
```diff
@@ -126,17 +129,91 @@ func BenchmarkTxCrud(b *testing.B) {
...
+	t.Log("Verifying update and find in same transaction")
+	user1.Name = "Jeyhun G."
+	user1.Email = "jeyhun111@mail.com"
+	user1.EmailConfirmed = !user1.EmailConfirmed
+	user1.NumFollowers++
+	user2.Name = "Jahangir G."
+	user2.Email = "jahangir111@mail.com"
+	user2.EmailConfirmed = !user2.EmailConfirmed
+	user2.NumFollowers--
+	tx, err = ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:179]: Same

Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 6]
```diff
@@ -126,17 +129,91 @@ func BenchmarkTxCrud(b *testing.B) {
...
+	assert := assert.New(t)
+	var err error
+
+	ctx := CokeAdminCtx
+	ds, _ := SetupDataStore("TestSingleTxCrud")
+	defer ds.Reset()
+	_, user1, user2 := SetupDbTables(ds)
+
+	txFetcher := authorizer.SimpleTransactionFetcher{}
+	tx, err := ds.Helper().GetDBTransaction(ctx, datastore.GetTableName(user1), user1)
```
Comment (YanhaoLi07) [inline pkg/datastore/datastore_with_txcontext_test.go:157]: Same

Please use ds.GetTransaction() to init transaction. ds.Helper().GetDBTransaction() (or ds.GetDBTransaction()) is used in every CRUD function to either extract the transaction pointer out of ctx, or init a brand new one

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 1 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 2 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 3 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 4 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 5 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 6 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Done

**Comment by gaurav-vmware** [comment]
@krishnamiriyala Please modify the MR description, this is a test case fix.

cc @YanhaoLi07

**Comment by krishnamiriyala** [comment]
> @krishnamiriyala Please modify the MR description, this is a test case fix.
> 
> cc @YanhaoLi07  

Done

**Review by jeyhun — APPROVED** [review]

**Pull request merged**